src/main/java/com/redhat/devtools/lsp4ij/LSPIJUtils.java

```diff
@@ -595,12 +597,26 @@ public static int toOffset(int line, int character, @NotNull Document document)
 
         // Adjust position line/character according to this comment https://github.com/microsoft/vscode-languageserver-node/blob/ed3cd0f78c1495913bda7318ace2be7f968008af/textDocument/src/main.ts#L26
         if (line >= document.getLineCount()) {
-            // The line number is greater than the number of lines in a document, it defaults back to the number of lines in the document.
+            // The line number is greater than the number of lines in a document, it defaults to the last valid offset in the document.
             return document.getTextLength();
         } else if (line < 0) {
             // The line number is negative, it defaults to 0.
             return 0;
         }
+        // If the line number is 0 and the character is beyond the length of the first line, see if the language server
+        // for this document should treat such a situation as an offset-only position
+        else if ((line == 0) && (character > document.getLineEndOffset(line))) {
+            VirtualFile virtualFile = LSPIJUtils.getFile(document);
+            Project project = virtualFile != null ? ProjectLocator.getInstance().guessProjectForFile(virtualFile) : null;
+            PsiFile file = project != null ? getPsiFile(virtualFile, project) : null;
+            if ((file != null) && LanguageServiceAccessor.getInstance(project).hasAny(
```

Comment:
LSPIJUtils#toOffset is used for a given language server, so we should not using LanguageServiceAccessor here.

I think to have a clean API, LSPClientFeatures should expose toOffset and toPosition. It will allow to override it and manage offset,position according the language server.

The toOffset should have a new parameter serverReportsOffsetOnlyPositions and the toOffset of LSPClientFeatures should call LSPIJUtils#toOffset with this flag.

Reply:
I'm not sure that I follow. Are you saying that you want every current call to `LSPIJUtils.toOffset()` to instead call `LSPClientFeatures.toOffset()` and similar with `toPosition()`? If so, the ripple effect of that change is going to be massive, and it's going to require getting the `LSPClientFeatures` all over the place where those calls currently exist. Apologies if I'm misunderstanding the guidance here...

Reply:
I do, however, understand that you're saying that `hasAny()` is evaluating for multiple language servers for file and can result in an incorrect answer when `toOffset()` is called for a specific language server for which there are multiple registered language servers. Let me ponder that...

Reply:
> I'm not sure that I follow. Are you saying that you want every current call to `LSPIJUtils.toOffset()` to instead call `LSPClientFeatures.toOffset()` and similar with `toPosition()`? If so, the ripple effect of that change is going to be massive, and it's going to require getting the `LSPClientFeatures` all over the place where those calls currently exist. Apologies if I'm misunderstanding the guidance here...

I agree that it is not a trivial refactoring, but imho we should do that like I did for uri.

Perhaps we should start that in an another PR to provide toOffset and toPosition in lsp client features. Instead of working with lsp client features we shoould work with an interface like I did for uri.

Having toOffset and toPosition in lsp client features will be powerfull

Reply:
Indeed let see that after the release but I ghink toOffset and toPosition must be an api that you could override to manage your custme case with pure offset.

Lsp4ij should follow lsp spec and not manage specific usecases. Having toOffset as api can be override on your side znd manage the pure offset.